### ISSUE TITLE
DataBuffer.Type.HALF in autoencoder broken?

### ISSUE THREAD

**Issue by workless** [deeplearning4j/deeplearning4j#3513]
When running the training in an autoencoder with half precision the score is infinity and the input layer is corrupted
With single precision the training doesn't show any weird behavior.
On the first iteration the input layer is a clear black/white, on the second iteration it gets noisy as the output, leading me to believe the ram was re-allocated.

expected: running faster with lower precision but still a somewhat useful result, the input layer doesn't change.

#### Version Information

Deeplearning4j version 0.8.0
gentoo/linux 4.4.49
nvidia-cuda-sdk-8.0.61
nvidia-drivers-378.13
gtx960/4gb
i7-6700HQ/16gb ram

  private static final int sSeed = 1234;
  private static final int sIterations = 1;
  private static final int sListenerFreq = sIterations / 20;
  private static final int **sInputOutputLayerSize** = 7863;
  private static final int **sFirstLayerSize** = 2000;
  private static final int **sSecondLayerSize** = 500;
  private static final int **sThirdLayerSize** = 250;
  private static final int **sCompressedLayerSize** = 100;
  private static final double sLearningRate = 0.01;

      MultiLayerConfiguration conf = new NeuralNetConfiguration.Builder()
        .seed(sSeed)
        .iterations(sIterations)
        .weightInit(WeightInit.XAVIER_UNIFORM)
        .learningRate(sLearningRate)
        .learningRateScoreBasedDecayRate(0.2)
        .biasInit(0.0)
        .optimizationAlgo(OptimizationAlgorithm.STOCHASTIC_GRADIENT_DESCENT)
        .list()
        .layer(0, new RBM.Builder().nIn(sInputOutputLayerSize).nOut(sFirstLayerSize).lossFunction(
          LossFunctions.LossFunction.KL_DIVERGENCE).build())
        .layer(1, new RBM.Builder().nIn(sFirstLayerSize).nOut(sSecondLayerSize).lossFunction(
          LossFunctions.LossFunction.KL_DIVERGENCE).build())
        .layer(2, new RBM.Builder().nIn(sSecondLayerSize).nOut(sThirdLayerSize).lossFunction(
          LossFunctions.LossFunction.KL_DIVERGENCE).build())
        .layer(3, new RBM.Builder().nIn(sThirdLayerSize).nOut(sCompressedLayerSize).lossFunction(
          LossFunctions.LossFunction.KL_DIVERGENCE).build())
        .layer(4, new RBM.Builder().nIn(sCompressedLayerSize).nOut(sThirdLayerSize).lossFunction(
          LossFunctions.LossFunction.KL_DIVERGENCE).build())
        .layer(5, new RBM.Builder().nIn(sThirdLayerSize).nOut(sSecondLayerSize).lossFunction(
          LossFunctions.LossFunction.KL_DIVERGENCE).build())
        .layer(6, new RBM.Builder().nIn(sSecondLayerSize).nOut(sFirstLayerSize).lossFunction(
          LossFunctions.LossFunction.KL_DIVERGENCE).build())
        .layer(7, new OutputLayer.Builder(LossFunctions.LossFunction.MSE).activation(Activation.SIGMOID).nIn(
          sFirstLayerSize).nOut(sInputOutputLayerSize).build())
        .pretrain(false).backprop(true)
        .build();

      model = new MultiLayerNetwork(conf);
      model.init();

**Comment by raver119** [comment]
What's the input?

**Comment by workless** [comment]
an array of 7863 floats, normalized to either 0 or 1 (that's why i was surprised to see a grey input nodes),
batch size is 1000
total data set size is 10000

**Comment by workless** [comment]
When saving and restoring the model it gets heavily corrupted: from a nice initialized model it becomes initialized with only zero, score gets from infinity (when model gets created) to NaN (after restore).

The grey input node may be due to a bug in the visualizer.

**Comment by raver119** [comment]
What you call "input node" in this case? Weights of first layer? Activations coming out of it?

**Comment by raver119** [comment]
P.S. That still sounds like precision issue.

**Comment by workless** [comment]
i confirm the input layer corruption was on my side, sorry for the false alarm.
by input layer i meant weight in the dataset.

saving and restoring a FP16 model using ModelSerializer.writeModel/restore seems to be broken:
before saving you can see noisy activations but after save&restore the network model seem to be reinitialized to 0.

I'll try different things, double check implementation and update the issue in a few days.

**Comment by raver119** [comment]
impossible, that stuff is covered with tests. there was bug in 0.6.0, but since that time it was fixed and tests were added.

So, please tell me more how you test that stuff, and how you check for weights values
